lib/elixir/test/elixir/calendar/naive_datetime_test.exs

```diff
@@ -348,4 +348,39 @@ defmodule NaiveDateTimeTest do
       assert catch_error(NaiveDateTime.to_time(~T[00:00:00.000000]))
     end
   end
+
+  describe "beginning_of_day/1" do
+    test "precision" do
+      assert NaiveDateTime.beginning_of_day(~N[2000-01-01 23:00:07.123]) ==
+               ~N[2000-01-01 00:00:00.000]
+
+      assert NaiveDateTime.beginning_of_day(~N[2000-01-01 23:00:07]) == ~N[2000-01-01 00:00:00]
+    end
+  end
+
+  describe "end_of_day/1" do
+    test "precision" do
+      assert NaiveDateTime.end_of_day(~N[2000-01-01 23:00:07.123]) == %NaiveDateTime{
+               calendar: Calendar.ISO,
+               day: 1,
+               hour: 23,
+               microsecond: {999_999, 3},
+               minute: 59,
+               month: 1,
+               second: 59,
+               year: 2000
+             }
+
+      assert NaiveDateTime.end_of_day(~N[2000-01-01 23:00:07]) == %NaiveDateTime{
+               calendar: Calendar.ISO,
+               day: 1,
+               hour: 23,
+               microsecond: {999_999, 0},
```

Comment:
the fact it's `{999_999, 0}` might lead to surprising behaviour:

```elixir
import ExUnit.Assertions
start = ~N[2023-01-01 00:00:00]
assert NaiveDateTime.end_of_day(start) == 
       start |> NaiveDateTime.add(1, :day) |> NaiveDateTime.add(-1, :second)
```

```
** (ExUnit.AssertionError)

Assertion with == failed
code:  assert NaiveDateTime.end_of_day(start) ==
              start |> NaiveDateTime.add(1, :day) |> NaiveDateTime.add(-1, :second)
left:  ~N[2023-01-01 23:59:59]
right: ~N[2023-01-01 23:59:59]
```

the difference is:

```elixir
iex> NaiveDateTime.end_of_day(start).microsecond
{999999, 0}

iex> start |> NaiveDateTime.add(1, :day) |> NaiveDateTime.add(-1, :second) |> Map.fetch!(:microsecond)
{0, 0}
```

We'd see similar behaviour for different precisions e.g. milliseconds.

I'm honestly not sure what should be the behaviour here.

Reply:
one way to look at it is we should be able to write the assertion under question using just sigils:

```elixir
assert NaiveDateTime.end_of_day(~N[2000-01-01 23:00:07]) == ~N[2000-01-01 23:59:59]
```

and the fact we can't is a reason to revisit this?

Reply:
I created a new issue. We should probably truncate it based on the precision: `{(10 ** precision) - 1, precision}`

Reply:
I just opened a PR to fix this #12450.